maven/rules.bzl

```
@@ -187,25 +189,44 @@ def _aggregate_dependency_info_impl(target, ctx):
     deps_all = deps + exports + runtime_deps
 
     maven_coordinates = find_maven_coordinates(target, tags)
-    dependency = None
+    dependencies = []
 
+    # depend via POM
     if maven_coordinates:
-        # depend via POM
-        dependency = struct(
+        dependencies = [struct(
             type = "pom",
             maven_coordinates = maven_coordinates
-        )
-    else:
+        )]
+    # include generic compilation_outputs in JAR
+    elif target[OutputGroupInfo].compilation_outputs:
+        source_jars = target[OutputGroupInfo]._source_jars.to_list()
         # include in the JAR
-        dependency = struct(
+        dependencies = [struct(
             type = "jar",
             class_jar = target[OutputGroupInfo].compilation_outputs.to_list()[0],
-            source_jar = target[OutputGroupInfo]._source_jars.to_list()[-1],
-        )
+            source_jar = source_jars[-1] if source_jars else None,
+        )]
+    # include runtime output jars
+    elif target[JavaInfo].runtime_output_jars:
```

I really like this solution - it makes perfect sense that `compilation_outputs` would be undefined for a `java_import` target, given that the JAR(s) are already compiled!

Can we maybe remove the `elif target[OutputGroupInfo].compilation_outputs` branch entirely? Are there any cases when `runtime_output_jars` would be undefined? I note you proposed `full_compile_jars` as an alternative, so that is an option to consider as well.

---

> Can we maybe remove the elif target[OutputGroupInfo].compilation_outputs branch entirely?

That's a really good question. The main reason I didn't remove it was because `target[OutputGroupInfo].compilation_outputs` seemed to fit the more generic case, given that it uses the standard `OutputGroupInfo` provider. Adding another branch to catch `JavaInfo` specific providers originally was intended to narrow down and catch the edge cases for rules that don't expose `compilation_outputs` as part of  `target[OutputGroupInfo]`.

> it makes sense that compilation_outputs would be undefined for a java_import target, given that the JAR(s) are already compiled!

That being said, I think maybe the using the `JavaInfo` provider might just be the more targeted/correct version of the `OutputGroupInfo` branch, and given that this is a rule for publishing Java artifacts to some Maven repository, it might make sense _just_ to have that branch. It is a bit more sophisticated in how it gathers the compile and source JARs, which would support some cases not previously supported.

I did a test to verify if we could remove the "generic" branch, and if that would even support bundling `java_import` and other `kt_jvm_library` targets and it does. My main apprehension if someone is depending on bundling targets that don't provide `JavaInfo`. But in that case, they are the edge case, so at the very least, we could swap the branches such that the more sophisticated handling is done before the generic catch-all.

Thoughts?

---

I'm also thinking we take a bold step and remove the branch with `compilation_outputs` on the grounds that
1. After some internal testing, we can't find anywhere where it's necessary
2. In its use case, it is probably a suboptimal approach

So if it does cause a regression for anyone, we should be able to write a fix for it that is better than the current code.

---

Awesome. Makes sense to me. I'll remove that branch and push it up!

---

Just one last thing:
```suggestion
    else:
```
If there is a bug in our implementation, this change will ensure we throw an error rather than just keeping silent about it.

---

Yeah, makes sense. I added an explicit `fail` branch to ensure that it's clear why it failed:

https://github.com/sugarmanz/bazel-distribution/blob/a694a8c2c1ebb95cd2e52f37843b1aed2eeba915/maven/rules.bzl#L209-L210

---

That's even better 👍